### PR TITLE
gtsam files

### PR DESCRIPTION
**Pull request by JaiXJM-BB**

### CODE REVIEW

**Comment by eleir9268** [comment]
Hey Jai, how complete is the gtsam port? I'm wondering if I can use it to verify the METIS port (dependency) which doesn't have any useful tests, or if theres still some work to do to get gtsam ready.

**Comment by JaiXJM-BB** [comment]
> Hey Jai, how complete is the gtsam port? I'm wondering if I can use it to verify the METIS port (dependency) which doesn't have any useful tests, or if theres still some work to do to get gtsam ready.

It is currently passing all tests on my VM & RPI4, I'm just ironing out some formatting stuff and adjusting changes to help get it accepted upstream. 
The version on github works but is currently locked into building all tests -> I will push changes fixing that and more later today, after which I will be merging changes to v4.2 from v4.1.1 today or tomorrow.

Since you need tests anyways you can use the version up now, but be aware the build process takes ~1-3 hours and tests take a while too

**Comment by JaiXJM-BB** [comment]
[Uploading gtsam_test_output.txt…]()
Test output, 8.0, aarch64le

**Comment by JaiXJM-BB** [comment]
[Uploading run_tests_2_out.txt…]()

New tests built & added to run; failures found.

Most seem to be within floating point error, however there are ordering tests which fully failed:

`(108/200) testOrdering
Not equal:
expected:
Position 0: 3, 2, 5, 0, 4, 1
actual:
Position 0: 5, 4, 2, 1, 0, 3
/home/............./gtsam/gtsam/inference/tests/testOrdering.cpp:281: Failure: "assert_equal(expected, actual)" 
`

`(175/200) testSymbolicBayesTree
Not equal:
expected:
Position 0: 3, 2, 5, 0, 4, 1
actual:
Position 0: 5, 4, 2, 1, 0, 3
  gtsam/gtsam/symbolic/tests/testSymbolicBayesTree.cpp:735: Failure: "assert_equal(Ordering(list_of(3)(2)(5)(0)(4)(1)), ordering)" 
`

**Comment by JaiXJM-BB** [comment]
[run_tests_3_out.txt](https://github.com/user-attachments/files/17983933/run_tests_3_out.txt)
Floating points still slightly different. May fix via a tolerance, but will probably reach out to maintainers via a PR.

**Comment by JaiXJM-BB** [comment]
> [run_tests_3_out.txt](https://github.com/user-attachments/files/17983933/run_tests_3_out.txt) Floating points still slightly different. May fix via a tolerance, but will probably reach out to maintainers via a PR.
> 
> Any further changes will likely be done to the actual GTSAM code and not to build-files.

@eleir9268 The sorting errors had to do with the default METIS version that gtsam uses - the same exception occured on apple machines. If you end up testing METIS with gtsam let me know if testOrdering or testSymbolicBayesTree have any weird behaviour

**Comment by JaiXJM-BB** [comment]
[Uploading run_tests_4_out.txt…]()
More tests, unstable portions now included

**Comment by JaiXJM-BB** [comment]
Broken Files:
[run_tests_2_out.txt](https://github.com/user-attachments/files/18013976/run_tests_2_out.txt)
[run_tests_3_out.txt](https://github.com/user-attachments/files/18013977/run_tests_3_out.txt)
[run_tests_4_out.txt](https://github.com/user-attachments/files/18013979/run_tests_4_out.txt)
[run_tests_5_out.txt](https://github.com/user-attachments/files/18013981/run_tests_5_out.txt)

**Comment by JaiXJM-BB** [comment]
[run_tests_out_6.txt](https://github.com/user-attachments/files/18171746/run_tests_out_6.txt)
[test_separator_out.txt](https://github.com/user-attachments/files/18171747/test_separator_out.txt)
Seems to be cutting at the mirror edge in a chain. very weird.

**Comment by JaiXJM-BB** [comment]
Ok so after investigation the failing test is partitioning a graph that looks like this:

(X0) ---1--- (X1) ---1--- (X2) ---20--- (X3) ---1--- (X4)
Where (XN) is node N, ---n--- is a link with weight n

Output is partitioned into two as such:
Partition "A": X3-X4
Partition "B": X0-X1-X2

However the expected is listed as:
Partition "A": X0-X1
Partition "B": X2-X3-X4

Trial 1: after changing the weight to be '20' on the X1-X2 link, the X2-X3 link is still the one cut for partition, which makes sense as partitioning should be mostly layout based. 

Trial 2: Changing the enumeration of links. They were originally labelled in-order from 0 to 4, swapped x1-x2 and x2-x3 so they are now added in order x0-x1 x2-x3 x1-x2 x3-x4. This expectedly also did not change the results.

Trial 3: Direction. Tried changing direction of links (since this is a directionless graph, should not matter). Expectedly did not change results.

Trial 4: Inverted graph in terms of adding stuff i.e. swapping from lowest first to highest for everything.
This did change the graph, into 1-0 and 4-3-2. So it seems that it just picks based on key order if there are multiple valid solutions.

Trial 5: Set key order to match expected results. Set the key order to 2, 3, 4, 0, 1. This does not change the graph in any way. Tests now pass. 
Trial 5 Control: To make sure that it isn't just spitting out the keys in some order, tried 2 0 3 1 4. Still worked. Likely based on which is the initial key as a tiebreaker.

So, I'm not convinced this is an issue. I think both are equally valid partitions of the graph, so it is just choosing between them arbitrarily. The more concerning this is : `***Warning: mcore memory was not fully freed when destroyed.`, which I will investigate next.

**Comment by JaiXJM-BB** [comment]
[run_tests_out_21.txt](https://github.com/user-attachments/files/18211685/run_tests_out_21.txt)

**Comment by eleir9268** [comment]
Do you still get a memory leak warning?

**Comment by JaiXJM-BB** [comment]
> Do you still get a memory leak warning?

My understanding of mcore is that warning actually means the memory wasn't freed prior to the tracker being freed - but it is freed when the tracker is. Since all memory is allocated and tied to a tracker, it should be fine. This is supported by the fact that this doesn't register as a failed test. I'll check again this afternoon to be safe though.

File: ports/gtsam/qnx.nto.toolchain.cmake
Comment (eleir9268) [inline ports/gtsam/qnx.nto.toolchain.cmake:24]: Should use q++, less likely to be surprised by a missing c++ library and AFAIK the only reason we weren't using it on our old svn repos is that cmake puts up an unrecognized compiler warning.

**Review by eleir9268 — APPROVED** [review]
Nice work!

**Pull request merged**
